Sound fx fixes

erwinbonsma opened the pull request:
This fixes several problems in the Sound FX implementation.

Firstly, irrespective how much time was still remaining for a pattern, it always filled the complete buffer. This typically meant that sounds lasted a bit longer than specified, which was especially noticeable for short patterns.

Secondly, when no more patterns remained, the entire buffer would be cleared via a memset.
This would typically wipe samples from the last pattern which still had to be played.
This was especially noticeable after increasing the buffer size and/or decreasing the sample rate. For the default buffer size and sample rate this effect was minimal, as during a single update period most of the buffer would be emptied (as 44100 / 25 ~= 2048).

Thirdly, the implementation could be made more efficient, in particular the inner loops that generate the samples.

One possible drawback of the revised code is that the original code did not use CPU when no sound effect was active (by relying on the full clear by the memset). That's not the case with the current implementation. It continuously writes zeros to the buffer, albeit quite efficiently.

Such an optimization can still be added, but requires a bit more book-keeping. In particular, when the sound effect is completely writter to the buffer, zeros need to be written to the buffer for several update cycles until the entire buffer is zeroed. It's however arguable if such an optimization is worth the effort. If a game uses sound effects, it should budget enough CPU cycles so that utilisation does not exceed 100% when sounds are playing. Therefore I did not (yet?) add it.

erwinbonsma committed:
Improve sound effect generation

The old implementation had several limitations:

Firstly, when a pattern still had some time remaining, it filled the complete buffer.
This typically meant that sounds lasted a bit longer which was especially noticeable for short patterns.

Secondly, when no more patterns remained, the entire buffer would be cleared via a memset.
This would typically wipe samples from the last pattern which still had to be played.
This was especially noticeable after increasing the buffer size and/or decreasing the sample rate.
For the default buffer size and sample rate this effect was minimal, as during a single update period most of the buffer would be emptied (as 44100 / 25 ~= 2048).

Thirdly, the implementation could be made more efficient, in particular the inner loops that generate the samples.

erwinbonsma committed:
Introduce separate define for Sound FX buffer size

Sorunome commented:
> That's not the case with the current implementation. It continuously writes zeros to the buffer, albeit quite efficiently.

Do you have any benchmarks? There are already a few programs stretching the cpu out a lot

Sorunome commented:
So yeah, thank you for the PR but maybe the suggested optimization is appropriate, as to not use that much CPU

erwinbonsma committed:
Stop writing zeroes to buffer once it is fully cleared

No need to waste these CPU cycles.
At the default sample rate, this reduces CPU usage by 2-3%, so that seems worth it.

erwinbonsma committed:
Minor fix to layout

erwinbonsma commented:
Thanks Sorunome for quickly looking into this. I have extended the code as requested. This reduces CPU usage by 2-3% so at the expense of one extra pointer, seems worth it. :-)

Sorunome commented:
Sorry it took so long to get back to this! It completely fell off the radar, thank you for emailing!

And, of course, thank you for the PR!
